sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/CosmosTableSchemaInferer.scala

```diff
@@ -5,122 +5,125 @@ package com.azure.cosmos.spark
-    ))
-
-    def inferSchema(inferredItems : Seq[ObjectNode]): StructType = {
-        if (inferredItems.isEmpty){
-            // No documents to infer from
-            StructType(Seq())
-        }
+  extends CosmosLoggingTrait {
 
-        // Create a unique map of all distinct properties from documents
-        val uniqueStructFields = inferredItems.foldLeft(Map.empty[String, StructField])({
-            case (map, item) => inferDataTypeFromObjectNode(item) match {
-                case Some(mappedList) =>
-                    map ++ mappedList.map(mappedItem => {
-                        if (map.contains(mappedItem._1) && map(mappedItem._1).dataType != mappedItem._2.dataType){
-                            // If 2 documents contain the same property name but different type, we default to String
-                            (mappedItem._1, StructField(mappedItem._1, StringType))
-                        }
-                        else{
-                            mappedItem
-                        }
-                    })
-            }
-        })
-
-        StructType(uniqueStructFields.valuesIterator.toSeq)
-    }
+  private[spark] val RAW_JSON_BODY_ATTRIBUTE_NAME = "_rawBody"
+  private[spark] val TIMESTAMP_ATTRIBUTE_NAME = "_ts"
+  private[spark] val ID_ATTRIBUTE_NAME = "id"
+  private[spark] val ETAG_ATTRIBUTE_NAME = "_etag"
+  private[spark] val PREVIOUS_RAW_JSON_BODY_ATTRIBUTE_NAME = "_previousRawBody"
+  private[spark] val TTL_EXPIRED_ATTRIBUTE_NAME = "_ttlExpired"
+  private[spark] val OPERATION_TYPE_ATTRIBUTE_NAME = "_operationType"
 
-    def inferSchema(client: CosmosAsyncClient,
-                    userConfig: Map[String, String]): StructType = {
-        val cosmosReadConfig = CosmosSchemaInferenceConfig.parseCosmosReadConfig(userConfig)
-        if (cosmosReadConfig.inferSchemaEnabled) {
-            val cosmosContainerConfig = CosmosContainerConfig.parseCosmosContainerConfig(userConfig)
-            val sourceContainer = client.getDatabase(cosmosContainerConfig.database).getContainer(cosmosContainerConfig.container)
-
-            val queryOptions = new CosmosQueryRequestOptions()
-            queryOptions.setMaxBufferedItemCount(cosmosReadConfig.inferSchemaSamplingSize)
-            val queryText = cosmosReadConfig.inferSchemaQuery match {
-                case None => s"select TOP ${cosmosReadConfig.inferSchemaSamplingSize} * from c"
-                case _ => cosmosReadConfig.inferSchemaQuery.get
-            }
-
-            val queryObservable =
-                sourceContainer.queryItems(queryText, queryOptions, classOf[ObjectNode])
-
-            val feedResponseList = queryObservable.byPage.collectList.block
-            inferSchema(feedResponseList.asScala.flatten(feedResponse => feedResponse.getResults.asScala))
-        } else {
-            defaultSchemaForInferenceDisabled
+  private[spark] def inferSchema(inferredItems: Seq[ObjectNode]): StructType = {
```

Comment:
If the object is already `private` why do we need to specify the `[spark]` on the method?

Reply:
because the "object" doesn't have visibility - the private above is for the ctor - all visibility is pushed to functions/methods/fields - private on the object just means the ctro is package private (private here means implicity private[spark]) - all members by default would still be public.
Anyway - given how confusing visibility scope is for Java/.Net developers when working in Scala I would strongly prefer explicitly listing the access modified - even when it is the default. Even in .Net I prefer adding it explicitly because it is not always obvious.